### ISSUE TITLE
Implement a Way to Provide a "Shadow Definition" of a Builtin Type

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4322]
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/181497846).
<hr />

Pivotal place holder for https://github.com/enso-org/enso/issues/1175
To flesh out with Marcin
(Not Basic Libraries but easier for JD to track!)

## Comments:
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-16):

**Today:** Started looking at the issue and figuring out what needs to be done It should be finished by 2022-03-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Continue digesting the requirements, will schedule a meeting with Marcin

**Additional Information:** Tentative end date based on how much I was told it will take (Enso Bot - Mar 16, 2022)
<hr />
A bit more context on "shadow definitions" (note to myself): https://github.com/enso-org/enso/blob/develop/docs/runtime-roadmap.md?plain=1#L214 (Hubert Plociniczak - Mar 17, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-17):

**Progress:** Investigating the current situation with builtins. Didn't spend much time on it due to other commitments. Will be catching up on Friday and Saturday It should be finished by 2022-03-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Meeting with Martin + discovery (Enso Bot - Mar 18, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-18):

**Progress:** More investigation into the current implementation. After meeting with Marcin, looking at Polyglot Builtin as a potential first candidate for replacement. On the side, played with benchmarks to make sure I didn't break stuff with default args. It should be finished by 2022-03-30. (Enso Bot - Mar 18, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-21):

**Progress:** Debugging internals. Finding a right place to extend definition scope to include Builtins (for now only focusing on Polyglot) It should be finished by 2022-03-30. (Enso Bot - Mar 22, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-22):

**Progress:** Extending definitions scope to include builtin types. Breaking a lot of internal assumptions on the way, fixing them and seeing how far the rabbit hole goes. It should be finished by 2022-03-30. (Enso Bot - Mar 22, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-23):

**Progress:** Not much progress since yesterday, had to revisit my approach for defining the scope. But have an idea how to continue. It should be finished by 2022-03-30. (Enso Bot - Mar 24, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-24):

**Progress:** Got some examples working which turned out to be easier than I thought. Will clean it up a bit for a PR to get early feedback and see if I'm on the right track or not. It should be finished by 2022-03-30. (Enso Bot - Mar 25, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the provided date (2022-03-25):

**Progress:** Slowly progressing through the problems mentioned in WIP PR. Mainly struggled with a sub-bug (feature?) where the resolution of Polyglot returns methods from the wrong scope (the one with `enso_project` rather than the one with Polyglot methods). Once I fix that we should be done. It should be finished by 2022-03-30. (Enso Bot - Mar 28, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-28):

**Progress:** Was cleaning up the PR for final review when I got stuck. Turns out having Builtin_Type Polyglot in Builtins and Polyglot methods in std library is a bit of a recipe for disaster when it comes to auto importing stuff. Was trying to get some quick hacks in, got frustrated, decided to continue the next day. It should be finished by 2022-03-30. (Enso Bot - Mar 29, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-29):

**Progress:** Decided to give up on full auto import for the incremental step. Benchmarks went fine, PR is ready for review. Now looking at actually handling Builtin_Type annotation, moving everything from Builtins to std library and then I will be able to bring back auto imports (changes in specs will explain the details, if anyone is interested). It should be finished by 2022-03-30. (Enso Bot - Mar 29, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-30):

**Progress:** Started moving non-Polyglot types out of Builtins to std lib. Fixing problems as I go. Had to spend some time to figure out how pattern matching is implemented with BranchNodes and how it affects Builtins. Looking good so far. It should be finished by 2022-03-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Moving on with other types from Builtins. `Any` will likely be the biggest pain. (Enso Bot - Mar 31, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-31):

**Summary:** There is **7 days** delay in implementation of the [Implement a Way to Provide a "Shadow Definition" of a Builtin Type (#181497846) task](https://www.pivotaltracker.com/story/show/181497846).
It **will cause 3 days delay** for the delivery of this weekly plan.

**Delay Cause:** The initial estimation was more of a guess, unfortunately, given my unfamiliarity with the engine. (Enso Bot - Apr 1, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-31):

**Progress:** Moved Java and Array types out of Builtins. Fixed what appears to be a bug (feature?) during desugaring of complex types - Atoms' annotations would get overridden by annotations from the complex type itself. Getting fairly confident the approach will work, just fixing any problems encountered on the way. It should be finished by 2022-04-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Moving more types out of Builtins (Enso Bot - Apr 1, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the provided date (2022-04-01):

**Progress:** The current approach required too many qualified names for complex types that share the name with the atom and the module (see Array.enso). Figured a way to make it work better before moving to other types. Result is visible in the current PR state. It should be finished by 2022-04-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Moving more types out of Builtins (Enso Bot - Apr 4, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-04):

**Progress:** Working on Boolean Builting type which is proving to be a bit tricky. Cleaned up the WIP PR to reflect the current state of progress. It should be finished by 2022-04-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Continue on Boolean and other Builtin types. (Enso Bot - Apr 5, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-05):

**Progress:** Moved Boolean out of builtins "as is", without attempting to provide a more meaningful type definition. Turned out I was stepping on something that Edward is partially attempting to address. Started moving Any until I realized there are issues when Any module is not directly or indirectly imported during the run - compiler has a number of places where it requires its presence, e.g. when resolving symbols. Will need to introduce the concept of required modules that always get loaded, even when not imported. It should be finished by 2022-04-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Continue on Any and other Builting types. (Enso Bot - Apr 6, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-06):

**Progress:** Got rid of runtime search for builtin methods, as suggested by Jaroslav. After realizing that I had a different understanding how Builtin_Types should be special, based on Marcin's feedback, spent some time figuring out how to allow for types that live in stdlib but nevertheless are known to compiler. It should be finished by 2022-04-06.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Will move Meta out of Builtins first after misunderstanding about Polyglot. Will continue researching on how to support BuiltinTypes in compiler. Deadline will obviously need to extend (Enso Bot - Apr 7, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-04-07):

**Summary:** There is **7 days** delay in implementation of the [Implement a Way to Provide a "Shadow Definition" of a Builtin Type (#181497846) task](https://www.pivotaltracker.com/story/show/181497846).
It **will cause 3 days delay** for the delivery of this weekly plan.

The basic case works, just need to figure out how to deal with BuiltinTypes

**Delay Cause:** Misunderstood one, important, bit of the ticket. Had to go back to the drawing table (Enso Bot - Apr 8, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-07):

**Progress:** Moved Meta out of Builtins to demonstrate that @Builtin_Method can be easily handled when we don't have to deal with BuiltinTypes in the compiler. Sadly I realized I had to go back to the drawing table with BuiltinTypes approach. Had some ideas but need to experiment first. It should be finished by 2022-04-13.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Trying different approach to handle specially builtin types in the compiler. It seems most of the corner cases I had were with Boolean or Any, so will start there. (Enso Bot - Apr 8, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the last Friday (2022-04-08):

**Progress:** As mentioned before, I started with a clean setup. Looks like there is only one corner case when trying to extend a builtin type with a method in a local module. At that point symbol lookup is getting confused. It should be finished by 2022-04-13.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Fix the corner case and move the rest of types. (Enso Bot - Apr 11, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-11):

**Progress:** Completed the move for Any and Boolean builtin types. Cleaned up the PR, ready for another round of review. Some discussions with Jaroslav re Engine infrastructure. It should be finished by 2022-04-13.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Addressing potential review, looking at the infrastructure of LSP vs Engine in preparation for the engine planning discussions. (Enso Bot - Apr 12, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-12):

**Progress:** Minor cleanups to the shadow definitions proposal PR, still waiting on feedback, looking at the infrastructure of LSP & engine to figure out how suggestions are being generated. Also, got fed up with formatting issues for Scala and Java, submitted early proposal to integrate checks to CI. Looking if the changes made to shadow definitions will allow easily to implement the next ticket in line #181499077. It should be finished by 2022-04-13.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Addressing potential review, continue switching between understanding suggestions in LSB & engine and investigating #181499077 (Enso Bot - Apr 13, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-13):

**Progress:** Addressing the review. Feedback from Marcin seems overall positive so seeing the light at the end of the tunnel. It should be finished by 2022-04-13.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Continue to address the review, moving the rest of builtin types, some additional benchmarking (Enso Bot - Apr 14, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-04-14):

**Summary:** There is **7 days** delay in implementation of the [Implement a Way to Provide a "Shadow Definition" of a Builtin Type (#181497846) task](https://www.pivotaltracker.com/story/show/181497846).
It **will cause 3 days delay** for the delivery of this weekly plan.

Runtime tests that rely on positions are super fragile and hard to debug

**Delay Cause:** It took a bit longer to come up with a clean solution for builtin types. Remaining time is dedicated to fixing tests that get broken every single time when moving individual types to stdlib. (Enso Bot - Apr 15, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-14):

**Progress:** Addressed the review. Moving other types is mostly mechanical but also touches a lot of tests which is a bit fragile. Also builtin methods in Number/Integer/Small_Integer/Big_Integer did not conform to the general approach so had to adapt the solution a bit. It should be finished by 2022-04-20.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Continue to move individual builtin types. (Enso Bot - Apr 15, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-19):

**Progress:** Continued moving other types. Mostly slowed down by adapting tests as I adapt types one by one. It should be finished by 2022-04-20.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Cleanup existing PR so that I don't have to deal with merge conflicts later. Will submit the PR with more than half of builtin types moved and deal with the rest in a separate PR. (Enso Bot - Apr 20, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-20):

**Progress:** Cleaning up the PR, fixed one minor bug that manifested itself when moving Nothing to stdlib. While doing last checks I realized I broke benchmarks somewhere on the way, which relied on the existing Builtins setup when invoked via polyglot api. It should be finished by 2022-04-20.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Fixing benchmarks which will unfortunately delay the task again. (Enso Bot - Apr 21, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-04-21):

**Summary:** There is **5 days** delay in implementation of the [Implement a Way to Provide a "Shadow Definition" of a Builtin Type (#181497846) task](https://www.pivotaltracker.com/story/show/181497846).
It **will cause 3 days delay** for the delivery of this weekly plan.

The PR had a couple of informal reviews so feeling fairly confident about the state of the ticket, despite the delay

**Delay Cause:** Encountered some issues with benchmarks that got broken somewhere on the way.

**Possible solutions:** Have another machine to continuously test changes 😉 (Enso Bot - Apr 22, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-21):

**Progress:** Working on benchmarks, got them back to where they were. More cleanups in preparation for the final review. It should be finished by 2022-04-25.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Updating some documentation to remove WIP from the PR (Enso Bot - Apr 22, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the provided date (2022-04-15):

**Progress:** Continued moving other types and fixing tests along the way. It should be finished by 2022-04-25.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Continue to move individual builtin types. (Enso Bot - Apr 22, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-04-22):

**Progress:** Moved the final builtins into stdlib. Mostly dealing with Numbers which had a different setup than other builtins and discovered a few existing bugs on the way. Most importantly the move is done. It should be finished by 2022-04-25.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181497846). Updating documentation, addressing PR review (Enso Bot - Apr 22, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-04-25):

**Progress:** Cleaned up the PR for the final review. It should be finished by 2022-04-25.

**Next Day:** Next day I will be working on the [##181988063 task](https://www.pivotaltracker.com/story/show/181988063). Addressing any review comments, and either 181988063 or the other sbt feature request (Enso Bot - Apr 26, 2022)
<hr />
